### PR TITLE
Narrow types via join / join-condition

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/staabm/phpstan-dba/issues/527

### CODE REVIEW

**Comment by staabm** [comment]
@gharlan could you do me a favour and verify the expected types?

File: tests/default/data/sql-ast-narrowing.php [hunk 1]
```diff
@@ -285,4 +285,34 @@ public function sum(PDO $pdo): void
...
+    public function joinNonNullable(PDO $pdo): void
+    {
+        $stmt = $pdo->query('SELECT adaid, akid from ada join ak on (adaid = eladaid)');
+        assertType('PDOStatement<array{adaid: int<-32768, 32767>, 0: int<-32768, 32767>, akid: int<-2147483648, 2147483647>, 1: int<-2147483648, 2147483647>}>', $stmt);
+
+        $stmt = $pdo->query('SELECT adaid, akid from ada inner join ak on (adaid = eladaid)');
+        assertType('PDOStatement<array{adaid: int<-32768, 32767>, 0: int<-32768, 32767>, akid: int<-2147483648, 2147483647>, 1: int<-2147483648, 2147483647>}>', $stmt);
+
+        $stmt = $pdo->query('SELECT adaid, akid from ada left join ak on (adaid = eladaid)');
+        assertType('PDOStatement<array{adaid: int<-32768, 32767>, 0: int<-32768, 32767>, akid: int<-2147483648, 2147483647>|null, 1: int<-2147483648, 2147483647>|null}>', $stmt);
```
Comment (staabm) [inline tests/default/data/sql-ast-narrowing.php:313]: rows not matched in the main result, will be filled up with null, therefore even a non-nullable column can get null on outer joins

File: tests/default/data/sql-ast-narrowing.php [hunk 2]
```diff
@@ -285,4 +285,34 @@ public function sum(PDO $pdo): void
         $stmt = $pdo->query('SELECT sum(c_double) as sum from typemix');
         assertType('PDOStatement<array{sum: float, 0: float}>', $stmt);
     }
+
+    public function joinNullable(PDO $pdo): void
+    {
+        $stmt = $pdo->query('SELECT adaid, eladaid from ada join ak on (adaid = eladaid)');
+        assertType('PDOStatement<array{adaid: int<-32768, 32767>, 0: int<-32768, 32767>, eladaid: int<-2147483648, 2147483647>, 1: int<-2147483648, 2147483647>}>', $stmt);
+
+        $stmt = $pdo->query('SELECT adaid, eladaid from ada inner join ak on (adaid = eladaid)');
```
Comment (staabm) [inline tests/default/data/sql-ast-narrowing.php:294]: on inner join, nullable columns get non nullable

**Review by gharlan — APPROVED** [review]
I think these expectations are correct 👍

File: src/SchemaReflection/Join.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
+<?php
+
+declare(strict_types=1);
+
+namespace staabm\PHPStanDba\SchemaReflection;
+
+/**
+ * @api
+ */
+final class Join
```
Comment (staabm) [inline src/SchemaReflection/Join.php:10]: not sure it makes sense to model joins here. maybe we should instead use `SqlFtw\Sql\Dml\TableReference\Join` instead, as we need the low level information

File: src/SqlAst/QueryScope.php [hunk 4]
```diff
@@ -106,10 +107,15 @@ public function getType($expression): Type
...
+            foreach ($this->joinedTables as $join) {
+                $joinedTable = $join->getTable();
+
                 foreach ($joinedTable->getColumns() as $column) {
                     if ($column->getName() === $expression->getName()) {
-                        return $column->getType();
+                        if ($join->getType() === Join::TYPE_OUTER) {
+                            return TypeCombinator::addNull($column->getType());
+                        }
+                        return TypeCombinator::removeNull($column->getType());
```
Comment (staabm) [inline src/SqlAst/QueryScope.php:118]: I think we should only add/remove NULLs for columns which are part of the join condition, but not regular columns selected from joined tables

Comment (staabm) [reply src/SqlAst/QueryScope.php:118]: before doing so, we may wait a bit for sqlftw stabilization as we would need a deep dive into the AST

grafik

**Pull request merged**